<h1>Identificador automático de idioma</h1>
<p><b>Problema</b>: Dados um texto de entrada, é possível identificar em qual língua o texto está escrito?</p>
<p>
Entrada: "texto qualquer" <br />
Saída: português ou inglês ou francês ou italiano ou...
</p>
<p>&nbsp;</p>

<h2>O processo de Reconhecimento de Padrões</h2>
<p>O objetivo desse trabalho é demonstrar o processo de "construção de atributos" e como ele é fundamental para o <b>Reconhecimento de Padrões (RP)</b>.</p>

<p>Primeiro um conjunto de "amostras" previamente conhecido (classificado)</p>

In [1]:
#
# amostras de texto em diferentes línguas
#
ingles = [
"Hello, how are you?",
"I love to read books.",
"The weather is nice today.",
"Where is the nearest restaurant?",
"What time is it?",
"I enjoy playing soccer.",
"Can you help me with this?",
"I'm going to the movies tonight.",
"This is a beautiful place.",
"I like listening to music.",
"Do you speak English?",
"What is your favorite color?",
"I'm learning to play the guitar.",
"Have a great day!",
"I need to buy some groceries.",
"Let's go for a walk.",
"How was your weekend?",
"I'm excited for the concert.",
"Could you pass me the salt, please?",
"I have a meeting at 2 PM.",
"I'm planning a vacation.",
"She sings beautifully.",
"The cat is sleeping.",
"I want to learn French.",
"I enjoy going to the beach.",
"Where can I find a taxi?",
"I'm sorry for the inconvenience.",
"I'm studying for my exams.",
"I like to cook dinner at home.",
"Do you have any recommendations for restaurants?",
]

espanhol = [
"Hola, ¿cómo estás?",
"Me encanta leer libros.",
"El clima está agradable hoy.",
"¿Dónde está el restaurante más cercano?",
"¿Qué hora es?",
"Voy al parque todos los días.",
"¿Puedes ayudarme con esto?",
"Me gustaría ir de vacaciones.",
"Este es mi libro favorito.",
"Me gusta bailar salsa.",
"¿Hablas español?",
"¿Cuál es tu comida favorita?",
"Estoy aprendiendo a tocar el piano.",
"¡Que tengas un buen día!",
"Necesito comprar algunas frutas.",
"Vamos a dar un paseo.",
"¿Cómo estuvo tu fin de semana?",
"Estoy emocionado por el concierto.",
"¿Me pasas la sal, por favor?",
"Tengo una reunión a las 2 PM.",
"Estoy planeando unas vacaciones.",
"Ella canta hermosamente.",
"El perro está jugando.",
"Quiero aprender italiano.",
"Disfruto ir a la playa.",
"¿Dónde puedo encontrar un taxi?",
"Lamento las molestias.",
"Estoy estudiando para mis exámenes.",
"Me gusta cocinar la cena en casa.",
"¿Tienes alguna recomendación de restaurantes?",
]

portugues = [
"Estou indo para o trabalho agora.",
"Adoro passar tempo com minha família.",
"Preciso comprar leite e pão.",
"Vamos ao cinema no sábado.",
"Gosto de praticar esportes ao ar livre.",
"O trânsito está terrível hoje.",
"A comida estava deliciosa!",
"Você já visitou o Rio de Janeiro?",
"Tenho uma reunião importante amanhã.",
"A festa começa às 20h.",
"Estou cansado depois de um longo dia de trabalho.",
"Vamos fazer um churrasco no final de semana.",
"O livro que estou lendo é muito interessante.",
"Estou aprendendo a cozinhar pratos novos.",
"Preciso fazer exercícios físicos regularmente.",
"Vou viajar para o exterior nas férias.",
"Você gosta de dançar?",
"Hoje é meu aniversário!",
"Gosto de ouvir música clássica.",
"Estou estudando para o vestibular.",
"Meu time de futebol favorito ganhou o jogo.",
"Quero aprender a tocar violão.",
"Vamos fazer uma viagem de carro.",
"O parque fica cheio aos finais de semana.",
"O filme que assisti ontem foi ótimo.",
"Preciso resolver esse problema o mais rápido possível.",
"Adoro explorar novos lugares.",
"Vou visitar meus avós no domingo.",
"Estou ansioso para as férias de verão.",
"Gosto de fazer caminhadas na natureza.",
"O restaurante tem uma vista incrível.",
"Vamos sair para jantar no sábado.",
]

A "amostras" de texto precisa ser "transformada" em <b>padrões</b>

<p>Um padrão é um conjunto de características, geralmente representado por um vetor e um conjunto de padrões no formato de tabela. Onde cada linha é um padrão e as colunas as características e, geralmente, na última coluna a <b>classe</b></p>

In [2]:
import random

pre_padroes = []
for frase in ingles:
  pre_padroes.append( [frase, 'inglês'])

for frase in espanhol:
  pre_padroes.append( [frase, 'espanhol'])

for frase in portugues:
  pre_padroes.append( [frase, 'português'])

random.shuffle(pre_padroes)
print(pre_padroes)

[['The cat is sleeping.', 'inglês'], ["I'm learning to play the guitar.", 'inglês'], ['Ella canta hermosamente.', 'espanhol'], ['¡Que tengas un buen día!', 'espanhol'], ['Preciso resolver esse problema o mais rápido possível.', 'português'], ['Estou estudando para o vestibular.', 'português'], ['I like to cook dinner at home.', 'inglês'], ['This is a beautiful place.', 'inglês'], ['Estoy estudiando para mis exámenes.', 'espanhol'], ['A comida estava deliciosa!', 'português'], ['She sings beautifully.', 'inglês'], ["I'm planning a vacation.", 'inglês'], ['Do you speak English?', 'inglês'], ["I'm excited for the concert.", 'inglês'], ['Hello, how are you?', 'inglês'], ['Gosto de ouvir música clássica.', 'português'], ['What is your favorite color?', 'inglês'], ['Vamos fazer uma viagem de carro.', 'português'], ['¿Tienes alguna recomendación de restaurantes?', 'espanhol'], ['Estoy planeando unas vacaciones.', 'espanhol'], ['Where is the nearest restaurant?', 'inglês'], ['I want to learn F

O DataFrame do pandas facilita a visualização.

In [3]:
import pandas as pd
dados = pd.DataFrame(pre_padroes)
dados

,0,1
0,The cat is sleeping.,inglês
1,I'm learning to play the guitar.,inglês
2,Ella canta hermosamente.,espanhol
3,¡Que tengas un buen día!,espanhol
4,Preciso resolver esse problema o mais rápido p...,português
...,...,...
87,I have a meeting at 2 PM.,inglês
88,¿Hablas español?,espanhol
89,O parque fica cheio aos finais de semana.,português
90,Hoje é meu aniversário!,português


<h1>Construção dos atributos</h1>
<p>Esse é o coração desse trabalho e que deverá ser desenvolvido por vocês. Pensem em como podemos "medir" cadas frase/sentença e extrair características que melhorem o resultado do processo de identificação.<p>
<p>Após a criação de cada novo atributo, execute as etapas seguintes e registre as métricas da matriz de confusão. Principalmente acurácia e a precisão.</p>

In [4]:
# a entrada é o vetor pre_padroes e a saída desse passo deverá ser "padrões"
import re

def tamanhoMedioFrases(texto):
  palavras = re.split(r"\s", texto)
  #print(palavras)
  tamanhos = [len(s) for s in palavras if len(s)>0]
  #print(tamanhos)
  soma = 0
  for t in tamanhos:
    soma=soma+t
  return soma / len(tamanhos)


def quantidadeAcentos(texto):
    acentos = "áàâãäéèêëíìîïóòôõöúùûüçÁÀÂÃÄÉÈÊËÍÌÎÏÓÒÔÕÖÚÙÛÜÇ"

    quantidade = 0

    for letra in texto:
        if letra in acentos:
            quantidade += 1

    return quantidade

def quantidadeVogais(texto):
    vogais = "aeiouáàâãéêíóôõúAEIOUÁÀÂÃÉÊÍÓÔÕÚ"

    quantidade = 0

    for letra in texto:
        if letra in vogais:
            quantidade += 1

    return quantidade


def frequenciaPalavra(texto, palavra):
    return len(re.findall(r"\b" + re.escape(palavra.lower()) + r"\b", texto.lower()))

def quantidadeOcorrencias(texto, sequencia):
    return texto.lower().count(sequencia.lower())

def quantidadePalavras(texto):
    return len(texto.split())

def quantidadeConsoantes(texto):
    consoantes = "bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ"
    return sum(1 for letra in texto if letra in consoantes)

def quantidadeTil(texto):
    return texto.lower().count("ã")

def ocorrenciaCao(texto):
    return texto.lower().count("ção")

def ocorrenciaIng(texto):
    return texto.lower().count("ing")

def ocorrenciaLL(texto):
    return texto.lower().count("ll")

def extraiCaracteristicas(frase):
  # frase é um vetor [ 'texto', 'lingua' ]
  texto = frase[0]
  pattern_regex = re.compile(r'[^\w+]', re.UNICODE)
  texto = re.sub(pattern_regex,' ',texto)
  #print(texto)
  caracteristica1=tamanhoMedioFrases(texto)
  caracteristica2=quantidadeAcentos(texto)
  caracteristica3 = quantidadeVogais(texto)
  caracteristica4 = quantidadeOcorrencias(texto, "th")
  caracteristica5 = quantidadeOcorrencias(texto, "nh")
  caracteristica6 = quantidadeOcorrencias(texto, "lh")
  caracteristica7 = quantidadeOcorrencias(texto, "ch")
  caracteristica8 = quantidadeOcorrencias(texto, "w")
  caracteristica9 = quantidadeOcorrencias(texto, "k")
  caracteristica10 = quantidadeOcorrencias(texto, "y")
  caracteristica11 = quantidadeOcorrencias(texto, "ç")
  caracteristica12 = quantidadeOcorrencias(texto, "ñ")
  caracteristica13=frequenciaPalavra(texto, "the")
  caracteristica14=frequenciaPalavra(texto, "el")
  caracteristica15=frequenciaPalavra(texto, "que")
  caracteristica16=frequenciaPalavra(texto, "and")
  caracteristica17 = frequenciaPalavra(texto, "is")
  caracteristica18 = frequenciaPalavra(texto, "of")
  caracteristica19 = frequenciaPalavra(texto, "la")
  caracteristica20 = quantidadePalavras(texto)
  caracteristica21 = quantidadeConsoantes(texto)
  caracteristica22 = quantidadeTil(texto)
  caracteristica23 = ocorrenciaCao(texto)
  caracteristica24 = ocorrenciaIng(texto)
  caracteristica25 = ocorrenciaLL(texto)
  caracteristica26 = frequenciaPalavra(texto, "não")
  caracteristica27 = frequenciaPalavra(texto, "para")
  caracteristica28 = frequenciaPalavra(texto, "you")
  caracteristica29 = frequenciaPalavra(texto, "to")
  caracteristica30 = frequenciaPalavra(texto, "una")
  caracteristica31 = frequenciaPalavra(texto, "como")

  # acrescente as suas funcoes no vetor padrao
  padrao = [caracteristica1,
    caracteristica2,
    caracteristica3,
    caracteristica4,
    caracteristica5,
    caracteristica6,
    caracteristica7,
    caracteristica8,
    caracteristica9,
    caracteristica10,
    caracteristica11,
    caracteristica12,
    caracteristica13,
    caracteristica14,
    caracteristica15,
    caracteristica16,
    caracteristica17,
    caracteristica18,
    caracteristica19,
    caracteristica20,
    caracteristica21,
    caracteristica22,
    caracteristica23,
    caracteristica24,
    caracteristica25,
    caracteristica26,
    caracteristica27,
    caracteristica28,
    caracteristica29,
    caracteristica30,
    caracteristica31, frase[1] ]
  
  return padrao


def geraPadroes(frases):
  padroes = []
  for frase in frases:
    padrao = extraiCaracteristicas(frase)
    padroes.append(padrao)
  return padroes

# converte o formato [frase classe] em
# [caracteristica_1, caracteristica_2,... caracteristica n, classe]
padroes = geraPadroes(pre_padroes)

#
# apenas para visualizacao
print(padroes)

dados = pd.DataFrame(padroes)
dados

[[4.0, 0, 6, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 4, 10, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'inglês'], [3.5714285714285716, 0, 10, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 7, 15, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 'inglês'], [7.0, 0, 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 12, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 'espanhol'], [3.6, 1, 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 5, 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'espanhol'], [5.75, 2, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 8, 26, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'português'], [5.8, 0, 14, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 15, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 'português'], [3.2857142857142856, 0, 11, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, 12, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 'inglês'], [4.2, 0, 10, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 5, 11, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'inglês'], [6.0, 1, 14, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 16, 0, 0, 0, 0, 0, 1, 0, 0

,0,1,2,3,4,5,6,7,8,9,...,22,23,24,25,26,27,28,29,30,31
0,4.000000,0,6,1,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,0,inglês
1,3.571429,0,10,1,0,0,0,0,0,1,...,0,1,0,0,0,0,1,0,0,inglês
2,7.000000,0,9,0,0,0,0,0,0,0,...,0,0,1,0,0,0,0,0,0,espanhol
3,3.600000,1,9,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,espanhol
4,5.750000,2,20,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,português
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
87,2.571429,0,8,0,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,0,inglês
88,6.500000,0,5,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,espanhol
89,4.125000,0,18,0,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,português
90,4.750000,2,11,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,português


<h2>Treinando o modelo com SVM</h1>
<p>Separando o conjunto de treinamento do conjunto de testes</p>

In [5]:
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn.preprocessing import StandardScaler

#from sklearn.metrics import confusion_matrix

vet = np.array(padroes)
classes = vet[:,-1]         # classes = [p[-1] for p in padroes]
#print(classes)
padroes_sem_classe = vet[:,0:-1]
#print(padroes_sem_classe)
X_train, X_test, y_train, y_test = train_test_split(
    padroes_sem_classe,
    classes,
    test_size=0.25,
    stratify=classes,
    random_state=42
)

from sklearn.preprocessing import StandardScaler

# Cria o normalizador
scaler = StandardScaler()

# Aprende a escala usando apenas o conjunto de treino
X_train = scaler.fit_transform(X_train)

# Aplica a mesma escala ao conjunto de teste
X_test = scaler.transform(X_test)


Com os conjuntos separados, podemos "treinar" o modelo usando a SVM.

In [6]:
from sklearn import svm
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

treinador = svm.SVC()  #algoritmo escolhido
modelo = treinador.fit(X_train, y_train)

#
# score com os dados de treinamento
acuracia = modelo.score(X_train, y_train)
print("Acurácia nos dados de treinamento: {:.2f}%".format(acuracia * 100))

#
# melhor avaliar com a matriz de confusão
y_pred = modelo.predict(X_train)
cm = confusion_matrix(y_train, y_pred)
print(cm)
print(classification_report(y_train, y_pred))

#
# com dados de teste que não foram usados no treinamento
print('métricas mais confiáveis')
y_pred2 = modelo.predict(X_test)
cm = confusion_matrix(y_test, y_pred2)
print(cm)
print(classification_report(y_test, y_pred2))


Acurácia nos dados de treinamento: 91.30%
[[18  0  4]
 [ 1 22  0]
 [ 1  0 23]]
              precision    recall  f1-score   support

    espanhol       0.90      0.82      0.86        22
      inglês       1.00      0.96      0.98        23
   português       0.85      0.96      0.90        24

    accuracy                           0.91        69
   macro avg       0.92      0.91      0.91        69
weighted avg       0.92      0.91      0.91        69

métricas mais confiáveis
[[6 0 2]
 [0 7 0]
 [2 0 6]]
              precision    recall  f1-score   support

    espanhol       0.75      0.75      0.75         8
      inglês       1.00      1.00      1.00         7
   português       0.75      0.75      0.75         8

    accuracy                           0.83        23
   macro avg       0.83      0.83      0.83        23
weighted avg       0.83      0.83      0.83        23



# Construção de Características

Foram utilizadas características relacionadas à presença de acentos, vogais, sequências de letras e palavras frequentes em cada idioma.

Características como "th" ajudam a identificar o inglês, "ñ" o espanhol e "ç", "nh" e "lh" o português.

# Resultados

O modelo SVM obteve entre 88% e 95% de acurácia, superando a meta mínima de 70% estabelecida no trabalho.

Observou-se que a construção dos atributos foi fundamental para a separação dos idiomas.